### PR TITLE
[JAX] Add `tpsp_resource` in the `MeshResource` map

### PR DESCRIPTION
**Pull request by phu0ngng**
# Description

This PR is needed for the upcoming CollectiveGEMM feature, as one needs to be able to verify if the tensor sequence parallelism is enabled instead of pure tensor parallelism.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: examples/jax/encoder/test_multigpu_encoder.py [hunk 1]
```diff
@@ -264,7 +264,7 @@ def train_and_evaluate(args):
     ) as mesh, te.fp8_autocast(
         enabled=args.use_fp8,
         fp8_recipe=fp8_recipe,
-        mesh_resource=te.MeshResource(DEVICE_DP_AXIS, None, None, None),
+        mesh_resource=te.MeshResource(dp_resource=DEVICE_DP_AXIS),
```
Comment (jberchtold-nvidia) [inline examples/jax/encoder/test_multigpu_encoder.py:267]: 👍 thanks for cleaning all these up to be clearer named args!

File: examples/jax/encoder/test_model_parallel_encoder.py [hunk 2]
```diff
@@ -267,7 +267,10 @@ def train_and_evaluate(args):
     ) as mesh, te.fp8_autocast(
         enabled=args.use_fp8,
         fp8_recipe=fp8_recipe,
-        mesh_resource=te.MeshResource(DEVICE_DP_AXIS, DEVICE_TP_AXIS, None, None),
+        mesh_resource=te.MeshResource(
+            dp_resource=DEVICE_DP_AXIS,
+            tpsp_resource=DEVICE_TP_AXIS,
```
Comment (jberchtold-nvidia) [inline examples/jax/encoder/test_model_parallel_encoder.py:272]: Are we re-assigning this TP axis to `tpsp` because this model parallel example is doing closer to TPSP and we're reserving TP usage for JAX's TP that shards layernorm's hidden dim?

File: transformer_engine/jax/sharding.py [hunk 3]
```diff
@@ -43,44 +43,44 @@ def _get_mesh_info(resource: str, mesh: jax.sharding.Mesh):
...
+        BATCH_AXES: gsr.fsdp_resource or gsr.dp_resource,
         SEQLEN_AXES: None,
-        SEQLEN_TP_AXES: gsr.tp_resource,
+        SEQLEN_TP_AXES: gsr.tpsp_resource,
         SEQLEN_CP_AXES: gsr.cp_resource,
-        HEAD_AXES: gsr.tp_resource,
+        HEAD_AXES: gsr.tp_resource or gsr.tpsp_resource,
         HIDDEN_AXES: None,
-        HIDDEN_TP_AXES: gsr.tp_resource,
+        HIDDEN_TP_AXES: gsr.tp_resource or gsr.tpsp_resource,
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/sharding.py]: I think in most settings, a user would set all MeshResource fields to non-None values, even if they're not using it in the current configuration. Should the checks here be changed from
```
gsr.tp_resource or gsr.tpsp_resource
```
to 
```
gsr.tp_resource if is_mesh_axis_active(gsr.tp_resource) else gsr.tpsp_resource
``` 
where 
```
def is_mesh_axis_active(axis_name: Optional[str]):
    physical_mesh = _PXLA_THREAD_RESOURCES.env.physical_mesh
    return axis_name is not None and physical_mesh.shape[axis_name] > 1
```

File: transformer_engine/jax/sharding.py [hunk 4]
```diff
@@ -303,7 +305,11 @@ def global_shard_guard(resource: MeshResource):
     old_resources = _GLOBAL_MESH_RESOURCE
     try:
         _GLOBAL_MESH_RESOURCE = resource
+        _validate_mesh_resource_configuration()
         yield
+    except Exception as e:
+        _GLOBAL_MESH_RESOURCE = old_resources
+        raise e
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/sharding.py]: I don't think we need this `except` block here. A try/finally without an except should still raise all errors through it

Tried this out in a small sample online here: https://www.online-python.com/ZztpMArsR1
```

try:
    raise ValueError('test')
finally:
    print('hello')
```
and the ValueError is still being raised at the outer scope. But if this is a different case in this example and I missed something, lmk. Thanks!

File: transformer_engine/jax/sharding.py [hunk 5]
```diff
@@ -351,52 +357,3 @@ def all_reduce_max_along_all_axes_except_PP(x: jnp.array, mesh: jax.sharding.Mes
         if axis != global_mesh_resource().pp_resource:
             x = lax_paral_op(x, jax.lax.pmax, axis, mesh)
     return x
-
-
-# Deprecating Items ---------------------------------------------------------------
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/sharding.py:356]: Thanks for removing unused code! 🎉

**Review by jberchtold-nvidia — COMMENTED** [review]
Overall LGTM, left a few small comments. Thanks for the PR!

File: transformer_engine/jax/sharding.py (hunk 3 above)
Comment (phu0ngng) [reply transformer_engine/jax/sharding.py]: Good finding! I updated it accordingly. Thanks

File: examples/jax/encoder/test_model_parallel_encoder.py (hunk 2 above)
Comment (phu0ngng) [reply examples/jax/encoder/test_model_parallel_encoder.py:272]: We update to use `tpsp_resource` as we always do TP + SP under the TP name so far in the tests and examples.
`tp_resource` now can be used for TP or TP + Hidden, but should not be used for TP + SP.

**Comment by phu0ngng** [comment]
\te-ci JAX L1

**Review by jberchtold-nvidia — APPROVED** [review]
LGTM, thanks!

**Review by jberchtold-nvidia — APPROVED** [review]
LGTM, thanks!
